File: javascript/libraries/shared/src/server/handlerbasic.ts
// NODE ONLY!

import express from 'express';
import { getFunctionFromRequest, getServiceFromRequest, RequestMessage, ResponseMessage } from '../common';
import { lookupServiceFunc, Registry } from './registry';

export function newBasicHandler(registry: Registry): express.RequestHandler {
  return async (req, res, next) => {
    try {
      // retrieve request message
      const reqMsg = req.body as RequestMessage;

      if (!reqMsg.address) {
        res.status(400).send('address not given');
        return;
      }

      if (reqMsg.final !== true) {
        res.status(400).send('request needs to be final');
        return;
      }

      try {
        // get and call func
        const [service, func] = lookupServiceFunc(registry, getServiceFromRequest(reqMsg), getFunctionFromRequest(reqMsg), 'basic');
        const resOutput = await func.call(service, reqMsg.input);

        // create response message
        const resMsg = <ResponseMessage>{
          id: reqMsg.id,
          output: resOutput,
          final: true,
        };

        // write response
        res.json(resMsg);
      }
      catch (err) {

        // create response message
        const resMsg = <ResponseMessage>{
          id: reqMsg.id,
          error: (err.message || err) + '',
          final: true,
        };

        // write response
        res.json(resMsg);
      }
    }
    catch (err) {

      // delegate to standard error handler
      next((err.message || err) + '');
    }
  };
}

File: go/libs/shared/server/handlerbasic.go
package server

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"

	common "github.com/assetninja/assetexchange/go/libs/shared/common"
)

func newBasicHandler(registry *Registry) func(res http.ResponseWriter, req *http.Request) {

	return func(res http.ResponseWriter, req *http.Request) {

		// error util
		reportError := func(message string, code int) {
			http.Error(res, message, code)
			log.Print(message)
		}

		// read request message
		reqMsg := common.RequestMessage{
			ID:      0,
			Address: nil,
			Input:   nil,
			Final:   true,
		}

		if err := json.NewDecoder(req.Body).Decode(&reqMsg); err != nil {
			reportError(
				fmt.Sprintf("could not parse http body: %v", err),
				http.StatusBadRequest)
			return
		}

		if reqMsg.Address == nil {
			reportError("address not given", http.StatusBadRequest)
			return
		}

		if reqMsg.Final != true {
			reportError("request needs to be final", http.StatusBadRequest)
			return
		}

		// call the function
		resOutput, resError := registry.callBasicFunc(reqMsg.GetService(), reqMsg.GetFunction(), reqMsg.Input)

		// create response message
		resMsg := common.ResponseMessage{
			ID:     reqMsg.ID,
			Output: resOutput,
			Error:  resError,
			Final:  true,
		}

		// write response
		if err := json.NewEncoder(res).Encode(resMsg); err != nil {
			reportError(
				fmt.Sprintf("could not generate response body: %v", err),
				http.StatusInternalServerError)
		}
	}
}
